src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,162 @@
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return find(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = find(index);
+        T oldValue = node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T oldValue = find(index).value;
+        unlink(find(index));
+        size--;
+        return oldValue;
     }
 
     @Override
     public boolean remove(T object) {
+        for (Node<T> node = head; node != null; node = node.next) {
+            if (object == node.value
+                    || object != null && object.equals(node.value)) {
+                unlink(node);
+                size--;
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bounds " + index);
+        }
+    }
+
+    private void checkSizeAdd(int index) {
+        if (!(index >= 0 && index <= size)) {
+            throw new IndexOutOfBoundsException("Can't add element to index " + index);
+        }
+    }
+
+    private Node<T> find(int index) {
+        if (index > (size >> 1)) {
+            return findFromTail(index);
+        }
+        return findFromHead(index);
+    }
+
+    private Node<T> findFromTail(int index) {
+        Node<T> node = tail;
+        for (int i = size - 1; i > index; i--) {
+            node = node.prev;
+        }
+        return node;
+    }
+
+    private Node<T> findFromHead(int index) {
+        Node<T> node = head;
+        for (int i = 0; i < index; i++) {
+            node = node.next;
+        }
+        return node;
+    }
+
+    private void link(T value, int index) {
+        Node<T> next = find(index);
+        Node<T> newNode = new Node<>(null, value, next);
+        if (size > 0 && index == 0) {
+            next.prev = newNode;
+            head = newNode;
+        } else {
+            newNode.prev = next.prev;
+            next.prev.next = newNode;
+            next.prev = newNode;
+        }
+    }
+
+    private void unlink(Node<T> node) {
+        if (node == head) {
+            if (size > 1) {
+                node.next.prev = null;
+                head = node.next;
+            }
+        } else if (node == tail) {
+            if (size > 1) {
+                node.prev.next = null;
+                tail = node.prev;
+            }
+        } else {
+            node.prev.next = node.next;
+            node.next.prev = node.prev;
```

Comment:
this implementation is so complex that you made a mistake :)

what will happen if `node == head` but `size !> 1`? 

exactly - this method will do nothing :)

if you got into `if` - you won't get into `else if` or `else`

try to refactor this method, make it simpler to avoid such mistakes